NordVPN improvements | Change up and down script and add item to DietPi-Banner

svh1985 opened the pull request:
**Status**:  Ready
- [x] Change up and down to run on route change
- [x] Add NordVPN entry to DietPi-Banner
- [x] Add function to NordVPN to pass status to DietPi-Banner 

Fixes: #3079

**Commit list/description**:
- DietPi-Nordvpn | Changed up to route-up and down to route-pre-down. This will ensure that the scripts will only run after the VPN successfully connected.

svh1985 committed:
Merge pull request #1 from MichaIng/dev

Pull latest changes into fork

svh1985 committed:
Update dietpi-nordvpn

Changed up to route-up and down to route-pre-down. This will ensure that the scripts will only run after the VPN succesfully connected.

svh1985 committed:
Update dietpi-banner

First commit.
Add DietPi-NordVPN status to banner.

svh1985 commented:
@MichaIng I noticed that the order of the banner items is tied to the array number. How do we handle an item that is added in the middle of array? I thought the NordVPN item should live at index 7. But I guess this messes up the settings file once we push the upgraded DietPi-Banner no?

svh1985 committed:
Update dietpi-banner

Placeholder added in DietPi-Banner for NordVPN

svh1985 committed:
First work on dietpi-nordvpn

Started work on a function in DietPi-NordVPN to get status from commandline. The retun now includes:
[  OK  ] DietPi-NordVPN | Root access verified.
[  OK  ] DietPi-NordVPN | RootFS R/W access verified.

Need advise to solve this.
Also set NORDVPN_SERVICE=1 manually because it does not sets the global var from the Init('status') function properly.

Banner item NordVPN calls DietPi-NordVPN to get status.

MichaIng commented:
@svh1985 
Many thanks for this. Looks good, will test later.

Btw the order of the banner entries does not depend on the banner index, but just on the oder on the code. So no need to change existing indices :).

svh1985 commented:
Ok good to know, I’ll then have to correct the indexes.

svh1985 committed:
Merge pull request #2 from MichaIng/dev

Get latest code

svh1985 committed:
Update dietpi-banner

Update to maintain compatibility.

svh1985 commented:
@MichaIng do you know the best way I can suppress the:
[  OK  ] DietPi-NordVPN | Root access verified.
[  OK  ] DietPi-NordVPN | RootFS R/W access verified.
It's now echoed in the banner.

svh1985 committed:
Update dietpi-nordvpn

Added service usage info that gets returned when wrong command has been passed as input.

svh1985 committed:
Update dietpi-nordvpn

Typo in usage info

MichaIng commented:
@svh1985 
Totally forgot about this. Humm... One way would be to write status to a temporary file and read it back from banner. But then using `dietpi-nordvpn status` from console is somehow awkward.

Okay lets do it like this:
- I anyway wanted to mute the root user and R/W check in case of success. IMO it is sufficient to do this silently and only print error/prompt if check failed.
- For status info, neither root permissions nor R/W access are required, thus both checks should be moved to right before the menu loop (Config file read must be done as well after those checks). Otherwise currently, when logging in with non-root user, banner would still print or even prompt error.

I was also thinking about how to avoid doubled code (optional for now):
- Connection state + in case RX/TX estimation can be done in a separate function. This function could produce related variables.
- When calling `G_INTERACTIVE=0 /DietPi/dietpi/misc/dietpi-nordvpn status`, all whiptail dialogs are skipped, questions are answered with "no". With this, "status" does not require an own Init() argument, but it can simply echo from within the existing G_WHIP_YESNO "else" part.
- By this, also the doubled NORDVPN_SERVICE check is obsolete, as long as Read_Settings is moved to in front of menu while loop.

I will remove the positive root user + r/w access prints now.

MichaIng commented:
@svh1985 
Done: https://github.com/MichaIng/DietPi/commit/1d25c2d8f48be97d106b01b315401322fb45529e
G_CHECKs succeed silently now.

svh1985 committed:
Merge pull request #3 from MichaIng/dev

Pull latest changes

svh1985 committed:
Update dietpi-nordvpn

Changed some status text and added colors to the status.

svh1985 commented:
@MichaIng thanks, it works like a charm. I also added colors to the status, what do you think?
![image](https://user-images.githubusercontent.com/5213210/64131783-38697000-cdcb-11e9-93f7-e03bb672519d.png)
![image](https://user-images.githubusercontent.com/5213210/64131819-82525600-cdcb-11e9-881e-3b301a2e6428.png)

MichaIng commented:
@svh1985 
Beautiful!! Yeah the colour makes totally sense for a VPN, so users will recognise quickly, if VPN is up or down. This is especially important as long as we don't have a killswitch integrated. Another task, most likely for v6.27.

svh1985 commented:
@MichaIng Do you want to merge or shall we implement your suggestions first? When do you want to release v6.26?

Kill switch would be nice! The Down script with dietpi-service stop qbittorrent is not very stable. I get multiple dietpi-service instance errors, not sure what goes wrong there. qbittorrent takes a long time to quite when having an active transfer maybe a openvpn down script timeout kills the dietpi-service script.

MichaIng commented:
@svh1985 
> Do you want to merge or shall we implement your suggestions first?

I will have a look later, probably do some fine tuning and then merge.

> The Down script with dietpi-service stop qbittorrent is not very stable.

Use `systemctl stop qbittorrent` instead. DietPi-Services is overkill here. It is mainly useful for either interactive calls where you need formatted output/error handling, or to start/stop all known services automatically in reasonable order, before maintenance tasks like installs, backups and such.

Just to verify:
```
time dietpi-services stop qbittorrent
G_USER_INPUTS=0 dietpi-services stop qbittorrent
```
- These both work from console, right?
- I wrapped the first into `time` so you can see how long it takes. Perhaps we can find out the timeout, if present, which would be good to know of course.

svh1985 commented:
I think that the OpenVPN down command does not want to wait for 1m30s and just kills the wait leaving dietpi-services in "not finished" state. 

The commands you wanted me to run (directly from prompt):
First command:
```
root@RockPi:~# time dietpi-services stop qbittorrent

 DietPi-Services
─────────────────────────────────────────────────────
 Mode: stop qbittorrent

[  OK  ] DietPi-Services | stop : qbittorrent

real	1m30.368s
user	0m0.660s
sys	0m1.476s
```
Second:
```
root@RockPi:~# G_USER_INPUTS=0 dietpi-services stop qbittorrent

 DietPi-Services
─────────────────────────────────────────────────────
 Mode: stop qbittorrent

[  OK  ] DietPi-Services | stop : qbittorrent
```

MichaIng commented:
@svh1985 
Hmm but this is not usual that qbittorrent takes 1.5 minutes to shutdown. Actually this is the timeout of the systemd unit, before it simply kills the process.
Do you have very much torrents active or something? Or does qbittorrent log contain some error message during shutdown/stopping it?

svh1985 commented:
@MichaIng Just one transfer. Ill check the log later and report it back once im home.

MichaIng committed:
v6.26

+ DietPi-NordVPN | Minor alignment
+ DietPi-NordVPN | DietPi scripts don't need to be executed with bash command, since they all contain a bash shebang.

MichaIng committed:
v6.26

+ DietPi-Banner | Fix output entry indices

MichaIng committed:
v6.26

+ DietPi-NordVPN | Do root user and rootfs R/W check only when menu is to be loaded, otherwise printing status would prompt error messages even that both is not required. This allows to load DietPi-Globals as well just for menu, which speeds up status print.
+ DietPi-NordVPN | Force non-interactive mode on status print, to skip any whiptail prompt
+ DietPi-NordVPN | Update WAN IP only on before first menu load, on connect, disconnect and refresh, to speed up all other navigation steps, which do not affect the connection state. This is due the WAN IP check can take up to two seconds, which adds a significant delay.
+ DietPi-NordVPN | Move RX/TX obtaining into an own function and remove DietPi-Globals dependency from it. This allows to share very most code between menu mode and status print.
+ DietPi-NordVPN | Further coding and wording enhancements

MichaIng commented:
@svh1985 
I made some changes, so that root user and rootfs R/W check is not done for status print. Otherwise the banner would throw an error message when logging in as non-root user.

Then I could not hold myself to redo the NordVPN install, configuration and network state/info, so that most code is now shared between status print and menu mode. And did some other coding enhancements, that are not related to your PR. Perfectionist here 😅.

I hope you are okay with that and I did not destroy anything. Will run tests now to see if banner print works as before my changes (where it worked like a charm already, as root user), then from my side it is ready to merge. Is there anything you still want to do about it (cause status was still WIP)?
_________
Two little fixes added, now testing passed from my side 👍.
EDIT: Ah nope, `G_WHIP_YESNO` is called before globals are loaded... hmm
EDIT2: Solved.

MichaIng committed:
v6.26

+ DietPi-NordVPN | Print connection test result to console as well, to stops prior process animation

MichaIng committed:
v6.26

+ DietPi-NordVPN | Print correct status text if service is not configured

MichaIng committed:
v6.26

+ DietPi-NordVPN | Now that globals are only loaded on menu mode, G_INTERACTIVE has no effect and G_WHIP_YESNO from Init() must only be called on menu mode.

MichaIng committed:
v6.26

+ DietPi-NordVPN | Tiny

MichaIng committed:
v6.26

+ DietPi-NordVPN | Resolve G_WHIP_YESNO being called before globals are loaded

svh1985 committed:
Merge pull request #4 from MichaIng/dev

Pull latest changes

svh1985 commented:
@MichaIng nice improvements! Everything worked during my testing. I'll just push one fix for a small menu alignment issue. 

I have to admit I now miss the connection state "Connected" in the dietpi-nordvpn menu. Do you mind if we put it back? I can push it along with the menu fix. Then I think the PR is ready to be merged 🎉

Like this:
![image](https://user-images.githubusercontent.com/5213210/64413887-aa270f80-d092-11e9-96b8-d29372b8af68.png)

svh1985 committed:
Update dietpi-nordvpn

Misallignment in menu, changed usage to traffic and added the connected state again.

MichaIng commented on dietpi/misc/dietpi-nordvpn:
```diff
@@ -155,34 +198,19 @@ _EOF_
 
-		Update_Wan_Ip
-
-		local text_status="Status:\n - WAN IP: $WAN_IP\n"
```
@svh1985
Perhaps this "Status:" line is needless?
Just have:
```
WAN IP      : <IP>
State       : <Not configured|Connected|Disconnected>
Traffic     : Sent = ... | Received = ... # If connected
Subscription: https://... # If not connected or configured 
```

svh1985 replied:
Agreed!

MichaIng committed:
v6.26

+ DietPi-NordVPN | 
+ DietPi-NordVPN |

MichaIng commented:
Accidental commit before filling its comment:
- Moved systemd unit to `/etc/systemd/system`, since `/lib/systemd/system` should be reserved for distribution package manager (APT). A DietPi-Patch will be added to move units on existing installs.
**EDIT: https://github.com/MichaIng/DietPi/commit/9e726b5bc2ceb1ff4a8c57c4f190d51e555cac56**
- Fixed Check_Connected loop when applying settings, so that a check is done as well after waiting the last second. Currently when index reaches $MAX_WAIT_FOR_CONNECTION, it is waited for another second, but no check is done afterwards, which renders this last second a waste of time.
- Reordered menu status text as discussed above.
- Minor coding

![nordvpn](https://user-images.githubusercontent.com/28480705/64487300-ed22e780-d238-11e9-941e-4212fd0634a0.png)

MichaIng commented:
Changelog entry: https://github.com/MichaIng/DietPi/commit/5d0f793de0a9220fa322151282a2cd60029c7718
